PsfLauncher/main.cpp

```diff
@@ -268,43 +277,233 @@ void LaunchMonitorInBackground(std::filesystem::path packageRoot, const wchar_t
+        auto err{ ::GetLastError() };
+
+        if (execInfo.ApplicationName)
+        {
+            ss << L"ERROR: Failed to create a process for " << execInfo.ApplicationName << " ";
+        }
+        else
+        {
+            std::wstring applicationNameForError;
+            applicationNameForError = execInfo.CommandLine.substr(0, execInfo.CommandLine.find(' '));
+            ss << L"ERROR: Failed to create a process for " << applicationNameForError << " ";
         }
+
+        ErrorInformation error(ss.str(), err);
+        return error;
     }
 }
-int __stdcall wWinMain(HINSTANCE, HINSTANCE, PWSTR args, int cmdShow)
+
+ErrorInformation StartWithShellExecute(std::filesystem::path packageRoot, std::filesystem::path exeName, std::wstring exeArgString, LPCWSTR dirStr, int cmdShow)
 {
-    return launcher_main(args, cmdShow);
+    // Non Exe case, use shell launching to pick up local FTA
+    auto nonexePath = packageRoot / exeName;
+
+    CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
+    SHELLEXECUTEINFO shex;
+    memset(&shex, 0, sizeof(SHELLEXECUTEINFO));
+    shex.cbSize = sizeof(shex);
+    shex.fMask = SEE_MASK_NOCLOSEPROCESS;
+    shex.hwnd = (HWND)NULL;
+    shex.lpVerb = NULL;  // just use the default action for the fta
+    shex.lpFile = nonexePath.c_str();
+    shex.lpParameters = exeArgString.c_str();
+    shex.lpDirectory = dirStr ? (packageRoot / dirStr).c_str() : nullptr;
+    shex.nShow = static_cast<WORD>(cmdShow);
+
+
+    Log("\tUsing Shell launch: %ls %ls", shex.lpFile, shex.lpParameters);
+    if (!ShellExecuteEx(&shex))
+    {
+        auto err{ ::GetLastError() };
+        std::wostringstream ss;
+        ss << L"ERROR: Failed to create detoured shell process";
+
+        ErrorInformation error(ss.str(), err);
+        return error;
+    }
+
+    ErrorInformation noError;
+    return noError;
 }
+
+static inline bool check_suffix_if(iwstring_view str, iwstring_view suffix)
+{
+    if ((str.length() >= suffix.length()) && (str.substr(str.length() - suffix.length()) == suffix))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+void Log(const char* fmt, ...)
+{
+    std::string str;
+    str.resize(256);
+
+    va_list args;
+    va_start(args, fmt);
+    std::size_t count = std::vsnprintf(str.data(), str.size() + 1, fmt, args);
+    assert(count >= 0);
+    va_end(args);
+
+    if (count > str.size())
+    {
+        str.resize(count);
+
+        va_list args2;
+        va_start(args2, fmt);
+        count = std::vsnprintf(str.data(), str.size() + 1, fmt, args2);
+        assert(count >= 0);
+        va_end(args2);
+    }
+
+    str.resize(count);
+    ::OutputDebugStringA(str.c_str());
+}
+void LogString(const char* name, const char* value)
+{
+    Log("\t%s=%s\n", name, value);
+}
+void LogStringW(const char* name, const wchar_t* value)
+{
+    Log("\t%s=%ls\n", name, value);
+}
+
+bool CheckIfPowerShellIsInstalled()
+{
+    HKEY registryHandle;

```

Comment:
>registryHandle [](start = 9, length = 14)

this is leaked.  use RAII type library (like WIL) to close the handle

Reply:
Oh, right.  Didn't close the handle.

WIL.  Are you taling about [this?](https://github.com/microsoft/wil)